Port VimL's Blob type - vim-patch:8.1.{0735,0736,0738,0741,0742,0755,0756,0757,0765,0793,0797,0798,0802,1022,1023,1671},8.2.{0121,0184,0404,0521,0829,1473,1866,2712}

seandewar opened the pull request:
Ref #15192.

There's still more blob patches out there, but I think these are (more-or-less) enough to have a good initial implementation that's relatively bug-free (unless I screwed up somewhere of course :smile:).

The commit messages have many of the details, but some things of note:
- Added Blob support for Nvim's `chansend()`, `msgpackparse()`, `msgpackdump()`.
- Blobs convert to Lua strings.
- Blobs convert to API strings.
- Blobs convert to JSON arrays of numeric byte values.
  - Unlike Vim, there's a space after each comma, which is consistent with how Nvim does it for Lists.
- Blobs convert to messagepack BIN strings.
- `decode_string()` now converts binary strings with NULs to Blobs.
  - Makes Lua strings with NULs convert to Blobs.
  - Makes msgpack BINs with NULs convert to Blobs via `msgpackparse()`.
  - Previously, these were converted to msgpack BIN special dicts. (Ref: #14801)
  - If this change causes other issues, it can be limited to Lua -> VimL only.
- Blob globals stored in the ShaDa file should restore properly.
  - Works by storing an additional element alongside the encoded value to differentiate it from a stored String; necessary as both encode to msgpack BIN.
  - `VAR_TYPE_BLOB` is used as the value of the extra element, but as nothing else uses additional elements for globals, we could _technically_ use any value right now (might want to think about future compatibility though :eyes:).

Things not addressed in this PR that could be in the future:
- I did not include a way to easily convert Lua strings without NULs to Blobs. Maybe that could be addressed in the future. (A way to convert a VimL String to a Blob could be useful, similiar to what vim/vim#4049 was trying to do?)
- The msgpack utilities runtime plugin may need some updates to fully support Blobs.

seandewar committed:
vim-patch:8.1.0735: cannot handle binary data

Problem:    Cannot handle binary data.
Solution:   Add the Blob type. (Yasuhiro Matsumoto, closes vim/vim#3638)
https://github.com/vim/vim/commit/6e5ea8d2a995b32bbc5972edc4f827b959f2702f

Nvim-specific Blob conversions are implemented in future commits.

Refactor write_blob() to use a FileDescriptor, as f_writefile() was
refactored to use one (does not apply to read_blob()).

Use var_check_lock() in f_add() for Blobs from v8.1.0897.

Add a modeline to test_blob.vim and fix some doc typos.

Include if_perl.txt's VIM::Blob() documentation. Interestingly, this
function already worked before this port, as it just returns a Blob
string literal, not an actual Blob object.

N/A patches for version.c:

vim-patch:8.1.0741: viminfo with Blob is not tested

Problem:    Viminfo with Blob is not tested.
Solution:   Extend the viminfo test.  Fix reading a blob.  Fixed storing a
            special variable value.
https://github.com/vim/vim/commit/8c8b8bb56c724cc1bfc3d8520eec33f2d399697c

vim-patch:8.1.1022: may use NULL pointer when out of memory

Problem:    May use NULL pointer when out of memory. (Coverity)
Solution:   Check for blob_alloc() returning NULL.
https://github.com/vim/vim/commit/e142a9467a7f6845a426d8db6efedf246d3c13ac

seandewar committed:
vim-patch:8.1.0738: using freed memory, for loop over blob leaks memory

Problem:    Using freed memory, for loop over blob leaks memory.
Solution:   Clear pointer after freeing memory.  Decrement reference count
            after for loop over blob.
https://github.com/vim/vim/commit/ecc8bc482ba601b9301a6c129c92a0d1f8527f72

seandewar committed:
vim-patch:8.1.1023: may use NULL pointer when indexing a blob

Problem:    May use NULL pointer when indexing a blob. (Coverity)
Solution:   Break out of loop after using index on blob
https://github.com/vim/vim/commit/61be376337b0374d55a4b1d8206b2ec87ca54252

seandewar committed:
vim-patch:8.1.0742: not all Blob operations are tested

Problem:    Not all Blob operations are tested.
Solution:   Add more testing for Blob.
https://github.com/vim/vim/commit/05500ece6282407f9f7227aaf564e24147326863

Test_readfile_binary is already ported.

seandewar committed:
vim-patch:8.1.0736: code for Blob not sufficiently tested

Problem:    Code for Blob not sufficiently tested.
Solution:   Add more tests.  Fix uncovered crash.  Add test_null_blob().
https://github.com/vim/vim/commit/c0f5a78c15b194f23bedb82e6825e34f481e6532

eval0 and ex_echo's emsg-specific changes have already been ported.
These tests uncover another crash that was fixed in v8.1.0738.

seandewar committed:
vim-patch:8.2.0521: crash when reading a blob fails

Problem:    Crash when reading a blob fails.
Solution:   Avoid keeping a pointer to a freed blob object. (Dominique Pelle,
            closes vim/vim#5890)  Adjust error messages.
https://github.com/vim/vim/commit/15352dc6ec43fd50cc3be4f4fd1ad74d5619da20

seandewar committed:
vim-patch:8.1.0756: copy() does not make a copy of a Blob

Problem:    copy() does not make a copy of a Blob.
Solution:   Make a copy.
https://github.com/vim/vim/commit/3d28b58c519c9fc3427587201423c74746cc219e

Replace vim_memsave() with xmemdup().

seandewar committed:
vim-patch:8.1.0755: error message for get() on a Blob with invalid index

Problem:    Error message for get() on a Blob with invalid index.
Solution:   Return an empty Blob, like get() on a List does.
https://github.com/vim/vim/commit/2ea773b468a1143214c2f12b91ab5e1e7abb4a14

seandewar committed:
vim-patch:8.2.0121: filter() and map() on blob don't work

Problem:    filter() and map() on blob don't work.
Solution:   Correct the code. (closes vim/vim#5483)
https://github.com/vim/vim/commit/49c57ce50019b667e5005ce1cfb8cdc2e48bf868

seandewar committed:
vim-patch:8.1.0765: string format of a Blob can't be parsed back

Problem:    String format of a Blob can't be parsed back.
Solution:   Use 0z format.
https://github.com/vim/vim/commit/4131fd5509b283e978e8c6161f09643b64719787

seandewar committed:
vim-patch:8.1.0793: incorrect error messages for functions that take a Blob

Problem:    Incorrect error messages for functions that now take a Blob
            argument.
Solution:   Adjust the error messages. (Dominique Pelle, closes vim/vim#3846)
https://github.com/vim/vim/commit/0d17f0d1c09fa6db306336695ba646c21ea24909

seandewar committed:
vim-patch:8.1.0797: error E898 is used twice

Problem:    Error E898 is used twice.
Solution:   Rename the Blob error to E899. (closes vim/vim#3853)
https://github.com/vim/vim/commit/bf821bccf18453b01d25bee53e4954b02a5dd0e6

seandewar commented on src/nvim/shada.c:
```diff
@@ -3937,15 +3944,28 @@ static ShaDaReadResult shada_read_next_item(ShaDaReadDef *const sd_reader,
                    unpacked.data.via.array.ptr[0].via.bin.size);
-      if (msgpack_to_vim(unpacked.data.via.array.ptr[1],
-                         &(entry->data.global_var.value)) == FAIL) {
+      SET_ADDITIONAL_ELEMENTS(unpacked.data.via.array, 2,
```
This feels meh.
Also, I need to check that this even works properly.

bfredl replied:
how does it work? will existing shada files still be correctly read to reproduce stored strings?

seandewar replied:
For blobs, I use the `global_var.additional_elements` list to store a dummy `VAR_TYPE_BLOB` with the blob.
(the exact value doesn't matter right now as `global_var.additional_elements` doesn't seem to be used for anything else yet)

This seems to work and existing shadas should be (and appear to be) fine.

bfredl commented on src/nvim/api/private/defs.h:
```diff
@@ -102,6 +107,7 @@ typedef enum {
   kObjectTypeInteger,
   kObjectTypeFloat,
   kObjectTypeString,
+  kObjectTypeBlob,
```
let's not do this. A `kObjectTypeString` is already 8-bit clean.

seandewar committed:
vim-patch:8.1.0757: not enough documentation for Blobs

Problem:    Not enough documentation for Blobs.
Solution:   Add a section about Blobs.
https://github.com/vim/vim/commit/d89682477cd01ec60edd6d88093b95b12e180127

Include doc changes for empty() from v7.4.1274.
Include some minor typo fixes.

bfredl commented:
we should support a BLOB in `f_chansend` as well (just send the bytes).

seandewar committed:
feat(f_chansend): support Blob data argument

seandewar committed:
feat(f_msgpackdump): support dumping to Blob

seandewar committed:
feat(f_msgpackparse): support parsing from Blob

Note that it is not possible for msgpack_unpack_next() and
msgpack_unpacker_next() to return MSGPACK_UNPACK_EXTRA_BYTES, so it
should be fine to abort() on that.

Lua 5.1 doesn't support string hex escapes (\xXX) like VimL does (though
LuaJIT does), so convert them to decimal escapes (\DDD) in tests.

seandewar committed:
vim-patch:8.1.0798: changing a blob while iterating over it works strangely

Problem:    Changing a blob while iterating over it works strangely.
Solution:   Make a copy of the Blob before iterating.
https://github.com/vim/vim/commit/dd29ea18050284526174b5685781469240f5bc4a

seandewar committed:
vim-patch:8.1.1671: copying a blob may result in it being locked

Problem:    Copying a blob may result in it being locked.
Solution:   Reset v_lock. (Ken Takata, closes vim/vim#4648)
https://github.com/vim/vim/commit/b7b9efbccfe17ee3f7b1bb877b7745d5bfbf0804

seandewar committed:
vim-patch:8.2.0184: blob test fails

Problem:    Blob test fails.
Solution:   Check for different error when float feature is missing.
https://github.com/vim/vim/commit/92be6e3f46120bb8e6c8fca0a7868a08df8b3345

Nvim always has the float feature, but include the changes to the test
anyway.

seandewar committed:
vim-patch:8.2.0829: filter() may give misleading error message

Problem:    filter() may give misleading error message.
Solution:   Also mention Blob as an allowed argument.
https://github.com/vim/vim/commit/fcb0b61d15f66f0e9116a6bc56d9d8105bb913cf

Rename Test_map_fails() to Test_map_filter_fails() from v8.2.0610 and
include the modeline.

seandewar committed:
fix(f_insert): partially port v8.2.0634

Fixes a crash in f_insert() when inserting into a NULL blob.
Include blob-related test changes and some other simple changes.

seandewar committed:
vim-patch:8.2.2712: memory leak when adding to a blob fails

Problem:    Memory leak when adding to a blob fails.
Solution:   Clear the second typval before returning.
https://github.com/vim/vim/commit/f2dd9cb9958962302d33fab8fe6439645a2d1e1b

seandewar committed:
fix(f_remove): partially port v8.2.2779

Fixes remove() copying one extra byte after the end of a Blob's buffer.

Can't be fully ported as the change is from blob_remove(), which hasn't
been ported yet.

seandewar committed:
vim-patch:8.1.0802: negative index doesn't work for Blob

Problem:    Negative index doesn't work for Blob.
Solution:   Make it work, add a test. (closes vim/vim#3856)
https://github.com/vim/vim/commit/a5be9b62480a6f338a72c01e57c9edd0bca8048b

Leave tv_blob_get()'s return type untouched.

seandewar committed:
feat(shada): restore Blob globals properly

As Strings and Blobs are encoded as msgpack BINs, the current ShaDa
implementation will restore global Blob variables as Strings (or msgpack
special dicts if they contain NULs).

Encode an additional element with Blob globals to differentiate them
from Strings so that we can restore them with the correct type.

Adjust variables_spec.lua's autotest() to also check for proper type.

seandewar committed:
test(oldtest): unskip existing tests that use Blobs

seandewar committed:
vim-patch:8.2.1473: items in a list given to :const can still be modified

Problem:    Items in a list given to :const can still be modified.
Solution:   Work like ":lockvar! name" but don't lock referenced items.
            Make locking a blob work.
https://github.com/vim/vim/commit/021bda56710d98c09a6b35610a476ab2dd8c58ad

seandewar committed:
fix(test/dumplog): tostring(rv) before formatting as string

For example, implicitly converting a table to a string works in LuaJIT,
but needs to be done explicitly with tostring() in Lua 5.1.

This can cause issues when testing a non-JIT build if eq(), for example,
fails with a table argument. E.g: eq({}, {1}) will not print the details
of the assertion failure, but will instead print a less helpful "string
expected, got table" error.

seandewar committed:
feat(decode_string): decode binary string with NULs to Blob

Strings that previously decoded into a msgpack special for representing
BINs with NULs now convert to Blobs. It shouldn't be possible to decode
into this special anymore after this change?

Notably, Lua strings with NULs now convert to Blobs when passed to VimL.

seandewar reviewed (commented):
Ref: #14801.

Before https://github.com/neovim/neovim/pull/15211/commits/a84aa3559b00c547fc4bb08af2b5a5fad02d9c38, Lua strings with NULs would convert to msgpack BIN special dicts as described in `:h msgpack-special-dict`. After this commit, they'll instead be converted to Blobs.

The way this is implemented affects all conversions from msgpack, which hopefully shouldn't be an issue as Blobs are encoded to msgpack as BINs anyway. However, if this does cause issues I can restrict these changes to Lua only.

seandewar committed:
fix(eval): partially port v8.2.3284

These were issues that I found while porting that I fixed upstream. :^)

Very little of the patch can be exactly ported as we're a bit behind on
dependant patches (we also can't use the exact :for emsg, as we don't
support iterating over Strings yet), so just translate the fixes as best
as we can for now.

Include latest relevant doc changes from:
- v8.1.0815
- v8.2.2658

seandewar committed:
doc(eval): include latest relevant Blob changes

Also includes some small relevant nearby non-Blob changes and typo
fixes.

Changes are included from:
- v8.1.0815
- v8.1.0846
- v8.1.1084
- v8.1.2326
- v8.2.1969
- d89682477c
- d09091d495
- 53f7fccc94

seandewar committed:
vim-patch:8.2.0404: writefile() error does not give a hint

Problem:    Writefile() error does not give a hint.
Solution:   Add remark about first argument.
https://github.com/vim/vim/commit/18a2b87ca27c378a555b20f14a284d2ce3511427

seandewar committed:
feat(json): convert Blobs to array of byte values

Similiar to how Vim does it, but to be consistent with how Nvim encodes
lists, add a space after every comma.

seandewar committed:
feat(api): convert Blobs to API strings

Note that these are not NUL-terminated; the API supports this.

seandewar committed:
feat(msgpack): convert Blobs to BIN strings

seandewar committed:
feat(nlua): convert Blobs to strings

seandewar commented:
I think this is more-or-less ready for some ~~roasting~~reviews. 👀

seandewar commented:
Hmm, I realize I should maybe consider converting API strings with NULs to Blobs in `object_to_vim()`, but that would mean checking the API string for early NULs. 🤔

i.e: `nvim__id(0z12.00.34)` loses access to the `0z34` bit in the resulting string (though it's still allocated for and copied).

bfredl commented:
@seandewar I'd say prefer to _not_ change behaviour around calling the API from vimL. if a plugin/config writer is using the neovim API, they can just as well use lua around any usecase concerning strings with NUL (this even applies to shared vim8/neovim plugins, if you are in a `has('nvim')` branch just as well as calling the API you can call into lua, even `v:lua` is valid syntax in vim8 as long as you never take the branch using it )

janlazo commented on src/nvim/eval.c:
```diff
@@ -2332,7 +2371,50 @@ static void set_var_lval(lval_T *lp, char_u *endp, typval_T *rettv,
+        }
+      } else {
+        bool error = false;
+        if (!error) {
```
This is dead code.
It is missing `tv_get_number_chk(rettv, &error)`.
Vim patch had the following:

```diff
+		val = (int)tv_get_number_chk(rettv, &error);
+		if (!error)
+		{
+		    garray_T *gap = &lp->ll_blob->bv_ga;
+
+		    // Allow for appending a byte.  Setting a byte beyond
+		    // the end is an error otherwise.
+		    if (lp->ll_n1 < gap->ga_len
+			    || (lp->ll_n1 == gap->ga_len
+				&& ga_grow(&lp->ll_blob->bv_ga, 1) == OK))
+		    {
+			blob_set(lp->ll_blob, lp->ll_n1, val);
+			if (lp->ll_n1 == gap->ga_len)
+			    ++gap->ga_len;
+		    }
+		    else
+			EMSG(_(e_invrange));
+		}
```

seandewar replied:
Oh wow, thanks for spotting that!

No idea what happened there :smile:

janlazo reviewed (changes requested):
Reviewed 1st commit only. Haven't checked typval macros because some are set to abort.

seandewar commented:
Thanks for taking a look!
Is it okay if I rebase right now? @janlazo

janlazo commented on runtime/doc/eval.txt:
```diff
@@ -3621,9 +3780,12 @@ diff_hlID({lnum}, {col})				*diff_hlID()*
-		items.  A Number is empty when its value is zero.  Special
-		variable is empty when it is |v:false| or |v:null|.
+		- A |List| or |Dictionary| is empty when it does not have any
+		  items.
```
```diff
+               - A |String| is empty when its length is zero.
```
is missing. Sync with Vim's doc for `empty()`, excluding unsupported types (ie. job).

janlazo commented on src/nvim/eval/typval.c:
```diff
@@ -2169,6 +2240,46 @@ void tv_dict_alloc_ret(typval_T *const ret_tv)
+    if (len > 0) {
+      to->vval.v_blob->bv_ga.ga_data
+          = xmemdup(from->vval.v_blob->bv_ga.ga_data, (size_t)len);
+      if (to->vval.v_blob->bv_ga.ga_data == NULL) {
```
`xmemdup` succeeds with a non-NULL pointer or triggers an abort.

janlazo commented on src/nvim/eval/typval.h:
```diff
@@ -711,6 +721,65 @@ static inline bool tv_dict_is_watched(const dict_T *const d)
+  return b->bv_ga.ga_len;
+}
+
+static inline char_u tv_blob_get(const blob_T *const b, int idx)
+  REAL_FATTR_ALWAYS_INLINE REAL_FATTR_NONNULL_ALL REAL_FATTR_WARN_UNUSED_RESULT;
+
```
Return type changed to `int`  on patch v8.1.0802.

janlazo replied:
Nvrmnd. Noticed the commit description.

seandewar replied:
I'll address this later :+1:

seandewar replied:
> 
> 
> Nvrmnd. Noticed the commit description.

Oh, apparently I had a good reason at the time then 🤣

janlazo replied:
Maybe `uint8_t` but leave it for later.

janlazo commented:
@seandewar Go ahead and rebase.

janlazo commented:
Enable the following tests (if they pass):
- Test_rename_fails
- Test_undofile_truncated

seandewar commented:
Enabled Test_undofile_truncated (enabled Test_swapfile_delete & Test_rename_fails previously)
They seem to pass :+1:

janlazo commented on src/nvim/testdir/test_const.vim:
```diff
@@ -244,18 +244,22 @@ func Test_const_with_eval_name()
+    call assert_fails("let d['foo'] = 'hello'", 'E741:')
+    call assert_fails("let d.foo = 'hello'", 'E741:')
     call assert_fails("let d['bar'] = 'hello'", 'E741:')
     call assert_fails("unlet d['foo']", 'E741:')
```
Missing

```diff
+
+    " Modifying list or dict item contents is OK.
+    let lvar = ['a', 'b']
+    let bvar = 0z1122
+    const l2 = [0, lvar, bvar]
+    let l2[1][0] = 'c'
+    let l2[2][1] = 0x33
+    call assert_equal([0, ['c', 'b'], 0z1133], l2)
+
+    const d2 = #{a: 0, b: lvar, c: 4}
+    let d2.b[1] = 'd'
```

janlazo reviewed (approved):
LGTM

janlazo commented:
I'm fine with merging this as is if the blob implementation for shada doesn't break existing behavior with non-blob code. I appreciate @seandewar work on this because this port isn't difficult to review, diverges from vim patches only when needed, and documented the rationale for the implementation differences with Vim.

I didn't review the encoding,decoding dance for msgpack and lua code. https://github.com/neovim/neovim/pull/15211#discussion_r678640728 should be resolved before merging.

janlazo commented:
Does the blob support for lua conflict with patch v8.1.1043 which adds `vim.blob`?

seandewar commented:
`vim.blob` seems to use userdata to represent blobs like it does with the other Vim types, which I presumed was not the direction Nvim was going with nlua?

Might need somebody to chime in on that statement though.

janlazo commented:
Port the following patches:
- 8.2.1866 (omit vim9 changes)

Can be separate PR:
- blob_remove and full port of v8.2.2779, v8.2.3284,
- 8.2.2343
- 8.2.2757 and related patches (does it actually need vim9 code?)
- 8.2.2780
- 8.2.2781
- 8.2.2782
- 8.2.2783
- 8.2.3438

seandewar committed:
vim-patch:8.2.1866: Vim9: appending to pushed blob gives wrong result

Problem:    Vim9: appending to pushed blob gives wrong result.
Solution:   Set ga_maxlen when copying a blob.
https://github.com/vim/vim/commit/66fa5fd54f550c0790d36c20124c49493b323bfa

Vim9script is N/A.

justinmk commented:
> `vim.blob` seems to use userdata to represent blobs like it does with the other Vim types, which I presumed was not the direction Nvim was going with nlua?

Compatibility with `if_lua` is a non-goal at this point, there is low (negative) leverage there.
